File: day1/python/puzzle1.py
def sum_first_last_digits(file_path):
    # Initialize sum to zero
    total_sum = 0
    # Open the file and read it line by line
    with open(file_path, 'r') as file:
        for line in file:
            # Initialize variables to store the first and last numbers
            first_digit = None
            last_digit = None
            final_num = None

            # Traverse the line to find the first and last numbers
            for char in line:
                if char.isdigit() and first_digit is None:
                    first_digit = char
                    break
            for char in reversed(line):
                if char.isdigit() and last_digit is None:
                    last_digit = char
                    break
            final_num = first_digit + last_digit            
            # Add the first and last numbers to the total sum
            if final_num is not None:
                total_sum += int(final_num)

    # Return the final sum
    return total_sum

def main():
    # Get the file path from the user
    file_path = input("Enter the path to your text file: ")

    # Calculate and print the sum
    result = sum_first_last_digits(file_path)
    print("Sum of first and last numbers:", result)

if __name__ == "__main__":
    main()

File: day1/golang/puzzle1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"unicode"
)

func sumFirstLastDigits(filePath string) int {
	totalSum := 0

	file, err := os.Open(filePath)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return totalSum
	}
	defer file.Close()

	absFilePath, err := filepath.Abs(filePath)
	if err != nil {
		fmt.Println("Error getting absolute file path:", err)
		return totalSum
	}
	fmt.Println("File Path:", absFilePath)

	reader := bufio.NewReader(file)

	for {
		line, _, err := reader.ReadLine()
		if err != nil {
			break
		}

		trimmedLine := strings.TrimFunc(string(line), func(r rune) bool {
			return !unicode.IsDigit(r)
		})

		if len(trimmedLine) > 0 {
			firstDigit := string(trimmedLine[0])
			lastDigit := string(trimmedLine[len(trimmedLine)-1])
			finalNum := firstDigit + lastDigit

			finalInt, err := strconv.Atoi(finalNum)
			if err != nil {
				fmt.Println("Error converting to integer:", err)
				continue
			}
			totalSum += finalInt
		}
	}

	return totalSum
}

func main() {
	fmt.Print("Enter the path to your text file: ")
	var filePath string
	_, err := fmt.Scanln(&filePath)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	result := sumFirstLastDigits(filePath)
	fmt.Println("Sum of first and last numbers:", result)
}
